Model/Service/Index.php

```diff
@@ -315,4 +336,105 @@ public function markAsInSyncProductByIdAndStore($productId, $storeId)
+        $lastPage = $collection->getLastPageNumber();
+        $curPage = 1;
+        do {
+            $collection->clear();
+            $collection->setCurPage($curPage);
+            /** @var Product $magentoProduct */
+            foreach ($collection as $magentoProduct) {
+                $key = array_search($magentoProduct->getId(), $uniqueIds);
+                if (is_numeric($key)) {
+                    unset($uniqueIds[$key]);
+                }
+            }
+            ++$curPage;
+        } while ($curPage <= $lastPage);
+
+        // Flag the rest of the ids as deleted
+        $deleted = $this->nostoIndexCollectionFactory->create()->markAsDeleted($uniqueIds, $store);
+        $this->logger->info(
+            sprintf(
+                'Marked %d records as deleted for store %s',
+                $deleted,
+                $store->getName()
+            )
+        );
+    }
+
+    /**
+     * Discontinues products in Nosto and removes indexed products from Nosto product index
+     *
+     * @param NostoIndexCollection $collection
+     * @param Store $store
+     * @return int number of deleted products
+     * @throws NostoException
+     */
+    public function handleProductDeletion(NostoIndexCollection $collection, Store $store)
+    {
+        if ($collection->getSize() === 0) {
+            return 0;
+        }
+        $totalDeleted = 0;
+        $account = $this->nostoHelperAccount->findAccount($store);
+        if ($account instanceof NostoSignupAccount === false) {
+            throw new NostoException(sprintf('Store view %s does not have Nosto installed', $store->getName()));
+        }
+        $collection->setPageSize(self::PRODUCT_DELETION_BATCH_SIZE);
+        $lastPage = $collection->getLastPageNumber();
+        $curPage = 1;
+        do {
+            $collection->clear();
+            $collection->setCurPage($curPage);
+            $ids = [];
+            /* @var $indexedProduct NostoProductIndex */
+            foreach ($collection as $indexedProduct) {
+                $ids[] = $indexedProduct->getId();
+            }
+            try {
+                $op = new DeleteProduct($account, $this->nostoHelperUrl->getActiveDomain($store));
+                $op->setResponseTimeout(30);
+                $op->setProductIds($ids);
+                $op->delete(); // @codingStandardsIgnoreLine
+                $rowsRemoved = $collection->deleteCurrentItemsByStore($store);
+                $totalDeleted += $rowsRemoved;
+            } catch (\Exception $e) {
+                $this->logger->exception($e);
+            }
+
+            ++$curPage;
+        } while ($curPage <= $lastPage);
+
+        return $totalDeleted;
+    }
+
+    /**
+     * Fetches deleted products from the product index, sends those to Nosto
+     * and deletes the deleted rows from database
+     *
+     * @param Store $store
+     * @return int
+     * @throws NostoException
+     */
+    public function purgeDeletedProducts(Store $store)
```

Comment:
Shouldn't this method be on the repository class?

Reply:
Let's do the refactoring later. The whole service is already quite bloated and will be most likely splitted into a few service class.

Reply:
👍